### DISCUSSION TITLE
Additional pkgPath column in HTML template

### DISCUSSION THREAD

**Discussion by JanMosigItemis** [aquasecurity/trivy#5388] [Ideas]
### Description

Hello everyone,

I'd like to suggest a minor addendum to the HTML template which makes the life of me and my colleagues easier. 

When a vulnerability has been found, I find it helpful if the report was to contain the name of the artifact that contains the vulnerability. E. g. if a vulnerability has been found inside a Java dependency - which is usually a jar file - it makes things much more easier to debug if the report would contain the jar's name.

I am already using this "feature" in my daily work. I have a fully functional example available and could easily provide a PR if you don't mind.

### Target

Container Image

### Scanner

Vulnerability

**Comment by DmitriyLewen** [comment]
Hello @JanMosigItemis 

I'm worried about long file paths. They can make the HTML table unreadable.
Can you send some examples? 

Regards, Dmitriy

**Reply by JanMosigItemis** [reply]
What kind of examples? HTML code or actual paths?

However, we could try to force line breaks by using a div of some sort I guess.

**Reply by DmitriyLewen** [reply]
i meant report in `html` format with long filepath.

**Reply by JanMosigItemis** [reply]
Ah ok. Will have a look.

**Reply by JanMosigItemis** [reply]
Well it creates a horizontal scroll bar. Same behavior as the link column. E.g.

```
<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
    <style>
      * {
        font-family: Arial, Helvetica, sans-serif;
      }
      h1 {
        text-align: center;
      }
      .group-header th {
        font-size: 200%;
      }
      .sub-header th {
        font-size: 150%;
      }
      table, th, td {
        border: 1px solid black;
        border-collapse: collapse;
        white-space: nowrap;
        padding: .3em;
      }
      table {
        margin: 0 auto;
      }
      .severity {
        text-align: center;
        font-weight: bold;
        color: #fafafa;
      }
      .severity-LOW .severity { background-color: #5fbb31; }
      .severity-MEDIUM .severity { background-color: #e9c600; }
      .severity-HIGH .severity { background-color: #ff8800; }
      .severity-CRITICAL .severity { background-color: #e40000; }
      .severity-UNKNOWN .severity { background-color: #747474; }
      .severity-LOW { background-color: #5fbb3160; }
      .severity-MEDIUM { background-color: #e9c60060; }
      .severity-HIGH { background-color: #ff880060; }
      .severity-CRITICAL { background-color: #e4000060; }
      .severity-UNKNOWN { background-color: #74747460; }
      table tr td:first-of-type {
        font-weight: bold;
      }
      .links a,
      .links[data-more-links=on] a {
        display: block;
      }
      .links[data-more-links=off] a:nth-of-type(1n+5) {
        display: none;
      }
      a.toggle-more-links { cursor: pointer; }
    </style>
    <title>Trivy Report</title>
    <script>
      window.onload = function() {
        document.querySelectorAll('td.links').forEach(function(linkCell) {
          var links = [].concat.apply([], linkCell.querySelectorAll('a'));
          [].sort.apply(links, function(a, b) {
            return a.href > b.href ? 1 : -1;
          });
          links.forEach(function(link, idx) {
            if (links.length > 3 && 3 === idx) {
              var toggleLink = document.createElement('a');
              toggleLink.innerText = "Toggle more links";
              toggleLink.href = "#toggleMore";
              toggleLink.setAttribute("class", "toggle-more-links");
              linkCell.appendChild(toggleLink);
            }
            linkCell.appendChild(link);
          });
        });
        document.querySelectorAll('a.toggle-more-links').forEach(function(toggleLink) {
          toggleLink.onclick = function() {
            var expanded = toggleLink.parentElement.getAttribute("data-more-links");
            toggleLink.parentElement.setAttribute("data-more-links", "on" === expanded ? "off" : "on");
            return false;
          };
        });
      };
    </script>
  </head>
  <body>
    <h1>Trivy Report</h1>
    <table>
      <tr class="group-header"><th colspan="7">Test</th></tr>
      <tr class="sub-header">
        <th>Vulnerability ID</th>
        <th>Found in</th>
        <th>Package</th>
        <th>Severity</th>
        <th>Installed Version</th>
        <th>Fixed Version</th>
        <th>Links</th>
      </tr>
      <tr class="severity-HIGH">
        <td>CVE-XYZ</td>
        <td class="pkg-name">/this/is/a/very/very/very/very/very/very/very/very/very/very/very/very/very/very/very/very/very/very/long/path.file</td>
        <td class="pkg-name">com.my.package</td>
        <td class="severity">HIGH</td>
        <td class="pkg-version">1</td>
        <td>2</td>
        <td class="links" data-more-links="off">
        <a href="about:blank" target="_blank">this is a realy realy realy realy realy realy realy realy realy realy realy realy realy realy realy realy long link</a>
        </td>
      </tr>
      <tr><th colspan="6">No Misconfigurations found</th></tr>
    </table>
  </body>
</html>

```

![image](https://github.com/aquasecurity/trivy/assets/16935994/ee5cd8b6-4917-4aed-814f-032b5adfec05)

**Reply by DmitriyLewen** [reply]
hm... thanks, for example.

I am not sure if 2 fields without length limit is good.
Perhaps we can add character limit:
if path length is greater than `limit` => use only filename
else => use file path.

**Reply by JanMosigItemis** [reply]
Well how about doing this in the CSS then:

```
table, th, td {
        border: 1px solid black;
        border-collapse: collapse;
        word-break: break-all;
        padding: .3em;
      }
```

Will result in
![image](https://github.com/aquasecurity/trivy/assets/16935994/3cebf83c-b2c1-4ccd-bb38-23290ef40343)

**Reply by DmitriyLewen** [reply]
Looks better!
But can we do this only for new column?
The package name or table headers don't look very good in a few lines.

**Reply by JanMosigItemis** [reply]
Well I guess we could, but it would look like this:

![image](https://github.com/aquasecurity/trivy/assets/16935994/4147f2b3-20a0-4a40-8731-083db85d59fd)

Notice how there are no line breaks in the Links column which causes horizontal scrolling.

How about applying the line breaks to non header cells only, like this:

![image](https://github.com/aquasecurity/trivy/assets/16935994/6d9485d2-14b4-4964-be4f-7bc226bdd5a5)

**Reply by DmitriyLewen** [reply]
@JanMosigItemis Thanks a lot for your examples!

Hm...
It seems first option (horizontal scrollbar) was best:
User can scroll to links (if needed), but there are no big cells with single number (like fixed version).

@knqyf263 wdyt?

**Comment by fatihtokus** [comment]
Hi @JanMosigItemis  ,

I have a trivy plugin(https://github.com/fatihtokus/scan2html) that creates a well-formated filterable/sortable html report for any scan result of trivy. The usage is very simple and I guess it does what you are looking for, please have a look:

Install the plugin: $ trivy plugin install github.com/fatihtokus/scan2html
$ trivy scan2html k8s --report=all cluster result.html

Regards!

![image](https://github.com/aquasecurity/trivy/assets/19647150/0bb37e03-d248-424c-8d08-a28336faee65)

**Reply by JanMosigItemis** [reply]
Thanks for the hint. Looks promissing. Problem is: We are using the official Trivy image here and must not use any third party software. So I guess we can't just use your plugin. However, maybe we could merge things into Trivy itself? How about this @DmitriyLewen?

**Reply by DmitriyLewen** [reply]
Perhaps we can add this plugin to our [ecosystem](https://aquasecurity.github.io/trivy/v0.46/ecosystem/) as community plugin.
@JanMosigItemis wdyt? Will this suit you?

I think integration of this plugin into Trivy is not possible. We can only add pkgPath to html template.

**Reply by JanMosigItemis** [reply]
Community plugin is better than nothing. Anyway, I'd be happy to enhance the html template nevertheless.

**Reply by fatihtokus** [reply]
I think trivy has a strong mechanism that allows the community to extend it by introducing new plugins. So I would suggest using/extending/contributing trivy plugins as well. Do not forget scan2html is opensource like trivy! 

@DmitriyLewen adding this plugin to the ecosystem is a very good idea, shall I create a pr for that?

Regards,
Fatih

**Reply by DmitriyLewen** [reply]
@knqyf263 Can you take part in this discussion?
What do you think about changes for `html` template and the addition of plugin created by @fatihtokus to docs in `ecosystem` section?

**Reply by itaysk** [reply]
Hi there, we are not really maintaining additional output formats in Trivy anymore (except the basic ones) as this tends to be very subjective and customizable topic. The solution will be to use template or a plugin like @fatihtokus suggested. We also have plans to make "output plugins" which will make discovering and using such custom formatters easier https://github.com/aquasecurity/trivy/issues/4860

**Reply by knqyf263** [reply]
@DmitriyLewen Sorry I missed the discussion.

>@DmitriyLewen adding this plugin to the ecosystem is a very good idea, shall I create a pr for that?

@fatihtokus Yes, please!
